ekm/slashing_protection_storage.go

```diff
@@ -0,0 +1,185 @@
+	RemoveHighestProposal(pubKey []byte) error
+
+	// SetVersion saves the version of the slashing protection DB
+	SetVersion(version string) error
+	// GetVersion returns the version of the slashing protection DB
+	GetVersion() (string, bool, error)
+}
+
+type spStorage struct {
+	db     basedb.Database
+	logger *zap.Logger // struct logger is used because core.Storage does not support passing a logger
+	lock   sync.RWMutex
+
+	prefix []byte
+}
+
+func (s *spStorage) SetVersion(version string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.db.Set([]byte(versionPrefix), nil, []byte(version))
+}
+
+func (s *spStorage) GetVersion() (string, bool, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	obj, found, err := s.db.Get([]byte(versionPrefix), nil)
+	if err != nil {
+		return "", found, errors.Wrap(err, "failed to get version")
+	}
+	if !found {
+		return "", found, nil
+	}
+
+	return string(obj.Value), found, nil
+}
+
+func NewSlashingProtectionStorage(db basedb.Database, logger *zap.Logger, prefix []byte) SpStorage {
+	return &spStorage{
+		db:     db,
+		logger: logger.Named(logging.NameSlashingProtectionStorage).Named(fmt.Sprintf("%sstorage", prefix)),
+		prefix: prefix,
+		lock:   sync.RWMutex{},
+	}
+}
+
+func (s *spStorage) objPrefix(obj string) []byte {
+	return append(s.prefix, []byte(obj)...)
+}
+
+func (s *spStorage) SaveHighestAttestation(pubKey []byte, attestation *phase0.AttestationData) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if pubKey == nil {
+		return errors.New("pubKey must not be nil")
+	}
+
+	if attestation == nil {
+		return errors.New("attestation data could not be nil")
+	}
+
+	data, err := attestation.MarshalSSZ()
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal attestation")
+	}
+
+	return s.db.Set(s.objPrefix(highestAttPrefix), pubKey, data)
+}
+
+func (s *spStorage) RetrieveHighestAttestation(pubKey []byte) (*phase0.AttestationData, bool, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if pubKey == nil {
+		return nil, false, errors.New("public key could not be nil")
+	}
+
+	// get wallet bytes
+	obj, found, err := s.db.Get(s.objPrefix(highestAttPrefix), pubKey)
+	if err != nil {
+		return nil, found, errors.Wrap(err, "could not get highest attestation from db")
+	}
+	if !found {
+		return nil, false, nil
+	}
+	if obj.Value == nil || len(obj.Value) == 0 {
+		return nil, found, errors.Wrap(err, "highest attestation value is empty")
+	}
+
+	// decode
+	ret := &phase0.AttestationData{}
+	if err := ret.UnmarshalSSZ(obj.Value); err != nil {
+		return nil, found, errors.Wrap(err, "could not unmarshal attestation data")
+	}
+	return ret, found, nil
+}
+
+func (s *spStorage) RemoveHighestAttestation(pubKey []byte) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.db.Delete(s.objPrefix(highestAttPrefix), pubKey)
+}
+
+func (s *spStorage) SaveHighestProposal(pubKey []byte, slot phase0.Slot) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
```

Comment:
I don't get why do we need mutex here, I'm assuming the DB is thread safe. anyway if we decide to use it anyway you should wrap only the critical section (database write/read) in mutex. the validations and marshaling shouldn't be serialized.

If there was a write here that is based on a read from the database, I would understand the mutex used wanting to lock these actions togteher. but I can't see anything like this here.

Reply:
This code was not introduced in this pull request but was moved from another file. The mutex usage here was from the original design for ensuring atomic operations with the database. Your point on limiting the scope of mutex lock is valid so we can discuss it further. @moshe-blox wdyt?